jax/_src/dtypes.py

```diff
@@ -412,7 +413,7 @@ def is_weakly_typed(x):
   try:
     return x.aval.weak_type
   except AttributeError:
-    return type(x) in _weak_types
+    return type(x) in _python_scalar_types
```

Comment:
Do you need the `any()` clause on L290 given it checks the same thing here? Maybe that's a performance optimization?

Reply:
No, in L290 it's checking if `val is typ`, wheras here it's checking if `type(val) == typ`. You have to be careful about these things because numpy dtype objects have overloaded equality to be annoyingly polymorphic (e.g. `np.dtype('float64') == float` evaluates to `True`). Here since `type(x)` will never be a `np.dtype` object, we don't have to use object identity, and we can use the more idiomatic object equality.

Reply:
:open_mouth:

Reply:
Should L292 be consistent with L410-411?
It might be better to add a util function like "test_membership_by_identity"?

Reply:
And what about L459?